Full controller smoothed parameters support.

MorghusDragon opened the pull request:
This adds support for smoothing parameters in the full controller feature.

A list of smoothed parameters can be provided, each with individual smoothing duration values. Alternatively you can specify `*`, which will enable smoothing for all float parameters.

I was struggling with the UI again but hopefully it's good enough. I was using the BlendShape field as inspiration for this.

![image](https://github.com/VRCFury/VRCFury/assets/140460527/d08b79ea-de8a-4803-8991-3fd4a57a2ca6)

Also, I didn't realize how janky it is to get the parameters from the serialized object. There's probably a better way to do it, please let me know!

```
The changes made in this contribution are free and unencumbered software
released into the public domain.

Anyone is free to copy, modify, publish, use, compile, sell, or
distribute this software, either in source code form or as a compiled
binary, for any purpose, commercial or non-commercial, and by any
means.

In jurisdictions that recognize copyright laws, the author or authors
of this software dedicate any and all copyright interest in the
software to the public domain. We make this dedication for the benefit
of the public at large and to the detriment of our heirs and
successors. We intend this dedication to be an overt act of
relinquishment in perpetuity of all present and future rights to this
software under copyright law.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.

For more information, please refer to <https://unlicense.org>
```

MorghusDragon commented:
I just realized I should actually get the parameter list from the controller and not from the VRC parameters object, that would make more sense.

SenkyDragon commented on com.vrcfury.vrcfury/Runtime/VF/Model/Feature.cs:
```diff
@@ -152,6 +153,13 @@ public class BindingRewrite {
+        public class SmoothParamEntry {
+            public string name;
+            public float smoothingDuration = 0.2f;
+            public bool ResetMePlease2;
```
when you add an item to a list, unity makes it a copy of the second-to-last item for some random reason. There's literally no way to tell it to make a fresh copy, so instead, VRCFuryEditorUtils.List looks for a property with this name, sets it to true after creating the new element, then using reflection, walks the entire plain object searching for this field with a true value, and when it finds it, it recreates a fresh copy of the class.

SenkyDragon commented on com.vrcfury.vrcfury/Editor/VF/Feature/FullControllerBuilder.cs:
```diff
@@ -253,6 +255,27 @@ public class FullControllerBuilder : FeatureBuilder<FullController> {
             from.RewriteParameters(RewriteParamName);
 
+            var smoothedParams = model.smoothedPrms.ToDictionary(
+                x => x.name == "*" ? x.name : RewriteParamName(x.name), 
```
I don't think we should support * as an option. It's almost certainly going to break things, and (admittedly), smoothing is probably not the most performant thing in the world, so we shouldn't just encourage its adoption for absolutely everything everywhere.

MorghusDragon commented on com.vrcfury.vrcfury/Editor/VF/Feature/FullControllerBuilder.cs:
```diff
@@ -253,6 +255,27 @@ public class FullControllerBuilder : FeatureBuilder<FullController> {
+                x => x.name == "*" ? x.name : RewriteParamName(x.name), 
+                x => x);
+            var fromParams = from.parameters.ToDictionary(x => x.name, x => x);
+            from.RewriteParameters(param => 
```
I was trying to make use of the `includeWrites: false` behavior here, if I combine with the call above I won't know if I do a write or not, unless I change the `RewriteParameters` method to include that information in the callback somehow. I agree having access to the original param name would be nice, especially since the smoothed layers wouldn't get doubled up `VF_` prefixes. 
How do you suggest keeping the `includeWrites: false` behavior and merging it with the call above?

SenkyDragon replied:
Fair enough. BTW: `includeWrites` was previously actually totally unused, and was introduced specifically to support this use case ;)

MorghusDragon commented:
Removed the wildcard and made it so it suggests all parameters from the controllers. Already smoothed parameters are not suggested again and it pops up a little message when there are no more parameters left to smooth.

MorghusDragon commented:
@SenkyDragon I think I addressed your concerns.

- I added a new `Smooth` method that keeps track if which parameters should be smoothed
- A new builder that actually applies the smoothing at a later step
- Support for smoothing not only the FX controller

MorghusDragon commented on com.vrcfury.vrcfury/Editor/VF/Feature/ParameterSmoothingBuilder.cs:
```diff
@@ -0,0 +1,50 @@
+            var raw = c.GetRaw();
+            var paramsMap = raw.parameters.ToDictionary(x => x.name, x => x);
+            var smoothedParamCache = new ConcurrentDictionary<string, string>();
+            raw.RewriteParameters(param => {
```
Just found another issue here, when the smoothing service adds a new layer, it gets detected by the `RewriteParameters` call as well and rewrites the input of the smoothing blend tree as well.

I suppose there are two ways to fix this:

1. Make `RewriteParameters` collect all data first and then iterate over them, or
2. Make it so you can control the name of the smoothed parameter and add the smoothing layers after rewriting them

SenkyDragon commented:
Can you try the full-controller-smoothing branch from origin and see if it still works properly for you?

MorghusDragon commented:
I just rebased and pushed again, had to resolve some conflicts. I tested it and seems to work for me now, what issues were you having?

SenkyDragon commented:
the login expanding into ControllerManager was scary. I pulled it back into FullController, moved the details into the advanced foldout, and added your credit.

SenkyDragon commented on com.vrcfury.vrcfury/Editor/VF/Utils/AnimatorControllerExtensions.cs:
```diff
@@ -28,6 +29,18 @@ public static class AnimatorControllerExtensions {
         public static void RewriteParameters(this AnimatorController c, Func<string, string> rewriteParamName, bool includeWrites = true) {
+            IImmutableSet<StateMachineBehaviour> behaviors = null;
+
+            // Collect data before calling the 'rewriteParamName' callback, which could add 
```
What can cause this to happen?

MorghusDragon replied:
Not sure if it's still the same after your changes but previously when adding the smoothing in the callback, the parameters of the smoothing layers would also get rewritten

SenkyDragon replied:
AH hmm, that is tricky. It's really cleaner for the smoothing to happen before RewriteParameters is called (so we can just do one pass over the param list and smooth the ones that need smoothed), but you're right that then it runs into this issue.
